### PR TITLE
Add GetoptMode parser setting and implementation

### LINKED ISSUE #91: MissingRequiredOptionError when specifying required Value after enumerable Options
<a href="https://github.com/SuperFXMaster"></img></a> **Issue by [SuperFXMaster](https://github.com/SuperFXMaster)**
_Tuesday Oct 11, 2016 at 19:24 GMT_
_Originally opened as https://github.com/gsscoder/commandline/issues/361_

----

The following options class:

``` C#
public class CommandLineOptions
{
    [Value(0, Required = true)]
    public string InputFileName { get; set; }

    [Option('o', "output")]
    public string OutputFileName { get; set; }

    [Option('i', "include", Separator = ',')]
    public IEnumerable<string> Included { get; set; }

    [Option('e', "exclude", Separator = ',')]
    public IEnumerable<string> Excluded { get; set; }
}
```

When provided to the default parser, with the commandline `app.exe --exclude=a,b InputFile.txt` results in a `MissingRequiredOptionError: A required value not bound to option name is missing`. I expected the library to accept `InputFileName = InputFile.txt` and `--exclude=a,b` as `Excluded = { "a", "b" }`. Is there something else required to make this work as I expect, or is it a bug?

### LINKED ISSUE #396: parsing fails when value matches name of list parameter
```csharp
class Options
{
    [Option("deploy", Separator = ',', = "Projects to deploy")]
    public IList<string> Deploys { get; set; }

    [Option("profile", Required = true, HelpText = "Profile to use when restoring and publishing")]
    public string Profile { get; set; }

    [Option("configure-profile", Required = true, HelpText = "Profile to use for Configure")]
    public string ConfigureProfile { get; set; }
}
```
Parsing fails if passed `--configure-profile deploy --profile local` with `Required option 'profile' is missing.`
I think it's intepreting the `deploy` value as an option.

### LINKED ISSUE #420: Parsing issues when option has separator attribute and when empty value provided... fails
```
 [Option('b', "BookIds", Required = false, HelpText = "List of Books", Separator = ',')]
        public IEnumerable<string> Bookids { get; set; }

[Option('v', "Verbose", Required = false, Hidden = true, HelpText = "Generate Book details", Default = "Yes")]
        public string All { get; set; }
```
ProcessBooks -b -v NO

------------
If option has separator attribute and when empty value provided... fails; it considers other argument as its values..

In this case; -b has value -v and NO;

### LINKED ISSUE #454: Wrong management of sequences and values
Consider following definition of options:
```
    class Options
    {
        [Option('c', "channels", Required = true, Separator = ':', HelpText = "Channel names")]
        public IEnumerable<string> Channels { get; set; }

        [Value(0,
               Required = true,
               MetaName = "file_path",
               HelpText = "Path of archive to be processed")]
        public string ArchivePath { get; set; }
    }
```
When invoking `app.exe -c chanA:chanB file.hdf5` the expected result would be 
```
Channels = {"chanA", "chanB"} 
ArchivePath = "file.hdf5"
```
Instead, I get a `CommandLine.MissingRequiredOptionError`

If I remove the `Required` attribute from the value definition, parsing succeeds but I get:
`Channels = {"chanA", "chanB", "file.hdf5"} `
And the expected `ArchivePath `value ends up in the `Channels `definitions (why? there is no ":" separator)

Am I doing something wrong?

### LINKED ISSUE #510: Using IEnumerable for Option is always 'Required'
[Option('a', "aa", Required = false, Separator = ',')]
        public IEnumerable<string> A { get; set; }
[Option('b', "bb", Required = false)]
        public string B { get; set; }
[Value(0, Required = true)]
        public string C { get; set; }

For some reasons, if I only specify Option A and C, I will get the following error:

ERROR(S):
  A required value not bound to option name is missing.
  Error setting value to option 'a, aa': Check if Option or Value attribute values are set properly for the given type.

### LINKED ISSUE #601: Corner cases where getopt behavior is not mimiced: -- or --help as string values
The goal of this library is to [mimic the behavior of getopt](https://github.com/commandlineparser/commandline/issues/498#issuecomment-521757260), but there are a few corner cases where this library behaves differently than getopt would: in the handling of `--` or `--help` when they are the value of a string parameter.

## How getopt behaves

First, an illustration of how getopt works with the particular corner case I'm demonstrating. Let's look at the standard `gzip` and `gunzip` tools found with any Linux distribution. They take many options, but one of them is `--suffix` (or `-S` for short); this lets you specify a different suffix than the standard `.gz` for the compressed file. E.g. if you have a README.md file in the current directory, then `gzip -S .compressed README.md` will create a README.md.compressed file instead of README.md.gz.

Now, what do you think will happen if I run this command?

```plaintext
gzip -S -- README.md
```

The correct answer is that it will create a compressed file named `README.md--` in the current directory. Because the string `--` was specified immediately after an option that takes a string value, it was processed as the value for that option (the `--suffix` option), and so gzip created a file with a `--` suffix instead of `.gz`. Now look at these three examples:

```plaintext
1. gzip -- --help
2. gzip -S -- --help
3. gzip -S -- -- --help
```

What do you think these will do? Answer:

1. This will compress a file named `--help` in the current directory, and create a file named `--help.gz`.
2. This will print the help text, and do nothing else.
3. This will compress a file named `--help` in the current directory, and create a file named `--help--`.

Why did `gzip -S -- --help` print the help text? Because `--` was the value for the `-S` option, and so it was *not* treated as the "stop processing options now" marker. Then after the `-S` option was fully processed, the only remaining options were `--help`. Since `--help` was encountered, gzip displayed the help screen and did nothing else.

With the `gzip -S -- -- --help` line, OTOH, the first `--` became the value for the `-S` option. Then the second `--` was processed as an option, and had the "stop processing options now" meaning. So the `--help` text was treated as a value, and so it looked for a file named `--help` to compress. And since I specified that the compressed suffix should be `--`, the compressed file was named `--help--`.

## What CommandLine does

The current way CommandLine works is to [call a preprocessor function](https://github.com/commandlineparser/commandline/blob/3e43750c5e0c3a433c61d8b7df0bffb8cb9d76ec/src/CommandLine/Core/Tokenizer.cs#L46) to look for any `--` options and, if found, mark anything found after them as a value. But this would mean that in the `gzip -S -- --help` example, where the correct getopt-mimicing behavior would be to print the help text, CommandLine will instead return an error saying that `-S` needed a value and didn't get one.

This corner case actually shows a fundamental difference between the behavior of CommandLine and the behavior of getopt. CommandLine uses a tokenizer to parse the command-line arguments and decide, based on the presence of `-` or `--` at the front, to treat them as Name tokens or Value tokens. But if you [read the `getopt` source code](https://github.com/gcc-mirror/gcc/blob/master/libiberty/getopt.c) and figure out what it's actually doing, it's parsing one argument at a time, deciding whether that argument needs a value, and then if a value is needed, it swallows the next argument without further processing. Which is why you can pass `--` as the suffix in gzip, and it will happily accept that.

## What CommandLine should do

The tokenizer, instead of processing all the arguments at once and deciding whether they're names or values, should process each argument one at a time. Then the decision tree should look like:

- Is this option exactly `--` and EnableDashDash is true? Then stop processing; the rest of the arguments are all values.
- Is this option exactly `--` and EnableDashDash is false? Then it is the value `--`; continue processing the next argument.
- Does this option start with `--` and contain an equals sign? Then split it into two tokens, the part before the `=` is the name, and the part after the equals is the value. (Split at the first equals sign; any equals signs after that point would become part of the value).
- Does this option start with `--` and not contain an equals sign? Then we look at the list of option longnames that the tokenizer was given:
    - Name matches a boolean option: this is a name token. Resume tokenizing with the next argument (it is NOT swallowed).
    - **NEW FEATURE**: Name matches an int option and the option attribute has `AllowMultiple=true`: this is a name token. Resume tokenizing with the next argument (it is NOT swallowed). (This allows for things like `-v` or `--verbose` to be passed multiple times, like `-vvv`, which the parser will turn into Verbose=3 in the final options instance.)
    - Name matches an option that's neither of the two cases above (boolean or int with AllowMultiple): this is a name token, and the next argument is a value token *no matter what it is*. "Swallow" the next argument, and resume tokenizing with the argument after next.
- Does the option start with `-` and contain only letters that match shortnames? Split it into multiple shortnames. (I.e., `-lR` would become `Name("l"), Name("R")` if there are both `-l` and `-R` options).
- Does the option start with `-` and its first letter matches a shortname, but the rest does not? Split it into first letter & rest, and that's two tokens: Name(first letter) and Value(rest).
- Does the option start with `-` and have only one letter? Then it's a shortname, and we look at the type of the option with that shortname:
    - As above, if boolean, then don't swallow the next argument.
    - NEW FEATURE: As above, if int with AllowMultiple, then don't swallow the next argument.
    - As above, if other type, then swallow the next argument (WHATEVER it is) and treat it as a value.

## Conclusion

Unfortunately, if the goal of getopt compatibility is to be achieved, a big rewrite of the guts of CommandLine's tokenizer and parser will be needed, so this is a big job. But if we want to mimic the behavior of getopt, then that's what will be needed. And the behavior I described above is how getopt works.

Also unfortunately, this is probably going to be a breaking change, so it might end up requiring a 3.0 version number. Because some people might be very surprised when `--stringoption --booloption` ends up being parsed with `--booloption` as the string value of `--stringoption`; they would probably have come to expect that to produce a `MissingValueOptionError` for `--stringoption`. But surprise or not, the correct way to handle that is for `--booloption` to be the string value of `--stringoption` in that example.

### LINKED ISSUE #617: Enumerable of Enum doesn't work if followed by Values
I have such options:
```csharp
[Option("fm",  Separator=',', Default = new[] { FMode.S })]  
public IEnumerable<FMode> Mode { get; set; }
	 
[Option('q')]
public bool q { get;set; }
	 
[Value(0)]
public IList<string> Files { get; set; }
```

If I run my application with `--fm D,C a.txt` arguments, it says that option 'fm' is defined with a bad format. If I put any option between 'fm' and the value list, for example, `--fm D,C -q a.txt`, it is parsed successfully. You can check it in a fiddle: https://dotnetfiddle.net/yt5L9K

### LINKED ISSUE #619: Order of options where there are list options and values
Hi.

The attached file contains the configuration to parse a number of options that consist of list of elements and finally a list of filenames (after the --).
The problem appear when the last option prior to the -- is a list based option. In that case the parser responds with: 'A required value not bound to option name is missing'

The following command line work with the provided C# configuration:
--modules ../utilities/x64/Debug,../auxtool/x64/Debug --outdir ./x64/Debug -- m_xfunit.f03 m_xfunit_assertion.f03

The following command line does not (with the error mentioned above)
--outdir ./x64/Debug --modules ../utilities/x64/Debug,../auxtool/x64/Debug -- m_xfunit.f03 m_xfunit_assertion.f03

Am I overlooking something or is this an issue or limitation in the software?

Regards,
FranHi.

The attached file contains the configuration to parse a number of options that consist of list of elements and finally a list of filenames (after the --).
The problem appear when the last option prior to the -- is a list based option. In that case the parser responds with: 'A required value not bound to option name is missing'

The following command line work with the provided C# configuration:
--modules ../utilities/x64/Debug,../auxtool/x64/Debug --outdir ./x64/Debug -- m_xfunit.f03 m_xfunit_assertion.f03

The following command line does not (with the error mentioned above)
--outdir ./x64/Debug --modules ../utilities/x64/Debug,../auxtool/x64/Debug -- m_xfunit.f03 m_xfunit_assertion.f03

Am I overlooking something or is this an issue or limitation in the software?

Regards,
Fran
[CommandLineOptions.cs.txt](https://github.com/commandlineparser/commandline/files/4508735/CommandLineOptions.cs.txt)

### LINKED ISSUE #631: Is it possible to use repeated non-option-argument options?
ssh, tar, and other common tools allow for multiple non-option-argument options and act on their total count. (Implementing this with getopt is pretty trivial).

e.g. `tar -cvf` and `tar -cvvf` produce different output.

Is there a way to do this with CommandLine? I did not see anything promising in the documentation or examples/source code I perused, but I may have missed something.

(This might just be related to #594, but the issue there seems to be with repeated options with _required_ option-arguments)

### LINKED ISSUE #687: Sequences with separators, if a separator is provided on the command line, should STOP at the separator
Consider this Options class:

```csharp
public class OptionsWithSeparator
{
        [Option('c', "categories", Separator = ',', HelpText = "Categories")]
        public IEnumerable<string> Categories { get; set; }

        [Value(0, HelpText = "File name")]
        public string Filename { get; set; }
}
```

What would you expect to get from the commandline `myprog -c one,two filename.txt`?

If you expect that to produce `Categories = ["one", "two"], Filename = "filename.txt"`, I agree. But that's not what happens. What it currently produces is `Categories = ["one", "two", "filename.txt"], Filename = null`. This is caused by the fact that IEnumerable options try to grab every subsequent non-option argument that follows them, and only stops once another option has been encountered (or, if EnableDashDash is true and you're using the upcoming 2.9 release, once a `--` argument has been encountered).

I believe that this is unintuitive behavior, and that what *should* happen with IEnumerable options is the following:

* If there is no separator given, grab all available non-option args
* If a separator is given, but the next non-option arg doesn't contain the separator, continue to grab all available non-option args
* If a separator is given, and the next non-option arg contains the separator, STOP and don't grab any further values (**this would be new behavior**)

Currently the code follows the first two of those rules, but not the third.

Here are bug reports that are ultimately caused by this overly-greedy grabbing of IEnumerables, which would have been solved by the behavior I listed above:

#91
#454
#510
#617
#619 (which can also be fixed by setting EnableDashDash = true, but it would *also* be fixed by the behavior I'm proposing here)

Clearly, users expect a separator, if provided, to confine the IEnumerable's "grabbing" to just the parameter containing the separator. I'll submit a PR soon to implement this behavior.

### PR DESCRIPTION
**Pull request by rmunn**
Turning on Getopt mode automatically turns on the EnableDashDash and AllowMultiInstance settings as well, but they can be disabled by explicitly setting them to false in the parser settings. This is essentially #607 rebased onto the develop branch, but with an option (defaulting to false) to turn it on. If the user doesn't set GetoptMode to true, then no behavior changes.

Note that two unit tests will fail until https://github.com/commandlineparser/commandline/pull/682 is merged, because they test for the same bug that #682 fixes.

This PR includes the FlagCounter feature from #607 as well; I've made the FlagCounter feature and the GetoptMode feature two separate commits so that if you want to use one but not the other, it's easy to cherry-pick the one you want to use. They don't depend on each other; in particular, FlagCounter works even without GetoptMode being true. But I've included both of them in this PR because GetoptMode without FlagCounter feels incomplete: people running in Getopt mode are going to expect to be able to do `-vvv` and get VerboseLevel=3.

My next PR will be a rebasing of #608 onto develop. Once that's done, if I'm sure that this PR is going to be merged, I'll start working on writing documentation of GetoptMode for the CLP wiki. In the meantime, the unit tests in GetoptParserTests.cs are pretty self-documenting.

Fixes #601.
Fixes #631.

I've also fixed #687 in this PR (in commit https://github.com/commandlineparser/commandline/pull/684/commits/47618e0ddf43a380baaa54484b188118b30d8a6c); I chose to incorporate that fix into this PR because GetoptTokenizer, introduced here, contains the same buggy ExplodeOptionList method as the original Tokenizer, and both of them need to be fixed to properly fix #687. In addition to #687, this PR now:

Fixes #619
Fixes #617
Fixes #510
Fixes #454
Fixes #420
Fixes #396
Fixes #91

### CODE REVIEW

**Comment by rmunn** [comment]
The failing unit tests are expected, and are fixed by #682. So #682 should be merged first, and then I'll rebase this PR on top of the new head of the `develop` branch so that the unit tests will pass.

**Comment by rmunn** [comment]
Just rebased on top of the new state of `develop` now that #682 has been merged.

**Comment by rmunn** [comment]
@moh-hassan - https://github.com/commandlineparser/commandline/pull/684/commits/47618e0ddf43a380baaa54484b188118b30d8a6c contains a fix for eight different bugs (listed in the PR description above). I've chosen to include it in this PR because GetoptTokenizer.cs (introduced in this PR) had the same bug, and needed the same fix, as the original Tokenizer.cs. (Because the buggy code is a method that I copied over from the original Tokenizer without modifying it). I thought about whether to submit that bugfix as a separate PR, but that would have caused *this* PR to get rather messy (this PR would then depend on another PR that had to be committed first, and it would have been rather complicated to get them merged in the right order, kind of like what happened with my original #607).

However, if you decide that you do **not** want to take this PR, please let me know and I'll do the work of separating out the Tokenizer bugfix from the rest of 47618e0ddf43a380baaa54484b188118b30d8a6c so that that bugfix, at least, can go in. But the simplest way by far is to just include that bugfix in this PR.

**Comment by moh-hassan** [comment]
Thanks @rmunn  for this Great PR. 
I agree for the bugfix  included in this PR.
ParserSettings.GetoptMode =true/false (false is default) is used. 
To enable future extensibility and avoid breaking change, It's planned to support other Commandline standard like Posix and forward slash in windows and enable the corresponding tokenizer:

- I suggest use enum like:
```cs
enum ParserMode
{
Gnu ,  //default
GetOpt , 
//Posix,   
//SingleDash , //like powershell , Unity engine, msbuild
//ForwardSlash  // windows 
 
}
```
you can rename the options.

- Can you provide specialized test cases that are unique to GetOptMode?
- Can you provide a wiki page how to use GetoptMode with code example?
- It's nice to share the common code between Tokenizer and GetoptTokenizer for maintenance.

**Comment by rmunn** [comment]
Hmm. The current parser mode isn't exactly GNU, because GNU getopt is what this PR is about emulating. I'm not quite sure what to call it, though: it's a weird mix of GNU behaviors and non-GNU behaviors. Perhaps "Default" mode would be a good name. And if there's ever a plan to change the default mode in the future (let's say there's widespread demand for Getopt mode to become the default in 3.x), then the current mode can be called "Legacy" mode when that breaking change is made, so that people who upgrade to 3.x but want to keep the 2.x behavior can just change "ParserMode = Default" to "ParserMode = Legacy" and the breaking change only forced them to change one line of code.

Or even better, we call the current mode "Legacy" right from the start, and *also* have an enum value called Default that is assigned to be the same as "Legacy" right now, but in 3.x might change to be assigned to some other mode. Yeah, I like that. That way people who use the default parser get the default mode, which might change in the future. And people who want to NOT have their current parser mode change in the future when we change the default can set it to "ParserMode = Legacy" right now, and we can guarantee that that won't change.

To answer your questions:

* This PR already includes specialized GetoptMode test cases, in the GetoptParserTests.cs file. Are they insufficient and you want to see more tests?
* I plan to work on the Wiki page next. I do want to wait until I know there will be no further changes needed to this PR and it will be merged as-is into the 2.9 release, because if (for example) I had already written up documentation for GetoptMode = true, I would have had to change it. So once you tell me, "Yes, this looks good, I plan to merge this without any more changes", I'll get started on the Wki documentation.
* The common code may not always stay common... but I guess if that happens we can copy the code from Tokenizer back into GetoptTokenizer and then make changes only to the GetoptTokenizer copy of the no-longer-common code. So I'll go ahead and deduplicate the common code now, and I can reverse that in the future if needed.

**Comment by moh-hassan** [comment]
@rmunn 
The term `Legacy` may mean it's becoming obsolete and don't address that the library support GNU standard (with customization).
I'll revert the last commit and use Getopt =true/false.

**Review by moh-hassan — APPROVED** [review]

**Comment by moh-hassan** [comment]
@rmunn 
Please, can you provide a wiki page for this feature with code example and guideline when to use getopt=true

**Comment by rmunn** [comment]
@moh-hassan There were some documentation improvements in that commit as well as the change to using an enum for ParserMode, so I'll submit a new PR with just those documentation changes so that they don't get lost by your reverting commit b7102d8.

BTW, if you decided to revert b7102d8, does that mean that you're giving up on the idea of a ParserMode enum in the future that would allow SingleDash or ForwardSlash options? Because if we release 2.9 with GetoptMode as a boolean, that will mean that we can't easily change to the ParserMode system in the future without breaking backwards compatibility for everyone who was using GetoptMode = true in their code. So instead of just reverting b7102d8, I wish you had just edited it to change the name Legacy to something you liked better. I picked Legacy because I think people *will* want Getopt mode to become the default in the future, but you could have easily edited it to call it Gnu if that's your preference. I think reverting that commit was a mistake, but as long as we fix that mistake before pushing the 2.9 release out the door, it won't become TOO big a problem. If we push 2.9 out the door as it stands now, then you're going to have backwards compatibility problems when you finally create the ParserMode enum that you asked for in https://github.com/commandlineparser/commandline/pull/684#issuecomment-677881462.

**Comment by rmunn** [comment]
@moh-hassan BTW, I'm going to start working on a Wiki page now.

**Comment by moh-hassan** [comment]
@rmunn 
The commit [b7102d8](https://github.com/commandlineparser/commandline/commit/b7102d89a90acd2d9356637ead5a0a64b58b131a) can be restored with changing the options of enum 
Do you agree for the next names?
```
CustomGetOpt  //default
GetOpt
```

**Comment by rmunn** [comment]
Yes, CustomGetopt would work. (The choices should be spelled Getopt and CustomGetopt, not GetOpt and CustomGetOpt, by the way: see [here](https://www.gnu.org/software/libc/manual/html_node/Getopt.html), for example, where the GNU documentation capitalizes the word getopt as Getopt, not as GetOpt).

**Comment by rmunn** [comment]
@moh-hassan Forgot to tag you on my reply, so here's a tag so you get a notification. As I said, the choices should be spelled CustomGetopt and Getopt (with a lowercase o, not uppercase O) but apart from that I'm okay with the name CustomGetopt instead of Legacy. It will need documentation explaining the difference between Getopt and CustomGetopt, but that's precisely what my wiki page is going to be. I'll follow the steps outlined in https://github.com/commandlineparser/commandline/issues/486 and submit patches. (Though if I'm going to be doing this a lot, I may end up asking for write access to the repo so that I can edit the wiki more directly and not have to bug you every time. But for now, I'll just submit a PR with my wiki edits).

**Comment by rmunn** [comment]
@moh-hassan - I have an even better idea. Instead of calling the current mode CustomGetopt or Legacy, we should call it Classic mode. Unlike Legacy, that doesn't imply that the mode is old or deprecated. Rather, calling it Classic mode implies that: 1) this is the current behavior, and 2) this is a mode that will be supported long into the future *even if we change the default behavior at some point*. So instead of calling it CustomGetopt, when you put commit b7102d8 back in, please call it Classic mode so that there will be less confusion in the long run.

**Comment by rmunn** [comment]
@moh-hassan I've opened https://github.com/commandlineparser/commandline/pull/690 so that you can just merge that in easily if you are happy with the name Classic.

**Comment by moh-hassan** [comment]
@rmunn 
Thanks for the suggestion.
It's better for the name to express the standard "Getopt" it implement, so developer don't confuse when set Parsermode to Classic. 
CustomGetopt means Tokenize with customization: e.g. 'AllowDashDash', 'IgnoreUnknowArguments', ...
So,kindly,  I'll merge the PR #690 with the name "CustomGetopt"  (with small letter o :)".

**Comment by rmunn** [comment]
@moh-hassan *Sigh*. That's a very bad decision that will just cause lots and lots of user confusion, but I can't stop you. The Getopt mode also allows the same amount of customization (you can turn AllowDashDash off if you want, IgnoreUnknownArguments works as designed, and so on), so haivng a CustomGetopt mode and a Getopt mode is going to do nothing but confuse people. I'm working on Wiki documentation explaining in detail the difference between Classic mode and Getopt mode, and I can certainly rename Classic to CustomGetopt in that documentation if you do carry through with that bad decision. But I again **URGE** you to reconsider and not make the mistake of calling these two modes something so similar that people will constantly get them confused. That's a mistake that will haunt the library forever if you make it now. Please don't do it.

P.S. Earlier I was okay with the name CustomGetopt, but that was before I started trying to explain the difference in documentation. Once I started to try to explain the difference I realized that the name CustomGetopt tells the user nothing about how it differs from Getopt, and so the similarity in the names is just going to confuse people for no reason. That's when I decided to call it Legacy, but Classic would have been even better which is why I went with Classic for #690.

**Comment by moh-hassan** [comment]
@rmunn 

What about these options:
```
GetoptParserV1   //default
GetoptParserV2
```
Both implement Getopt and no confusion .

**Comment by derrickstolee** [comment]
Hi, @rmunn. I'm interested in consuming a version of your package with this fix in https://github.com/microsoft/scalar. Do you have a timeline on the 2.9 release?

**Comment by rmunn** [comment]
@moh-hassan I have finally found time to work on CommandLineParser again. I like the GetoptParserV1 and V2 names you suggested in https://github.com/commandlineparser/commandline/pull/684#issuecomment-682600499, and I've implemented that change in #690. I'm also working on a wiki page explaining the difference between GetoptParserV1 and V2, and I'll submit it soon. Meanwhile, I think it might be a good idea to merge #690 before the 2.9 release goes out, so that the parser mode can be an enum instead of boolean GetoptMode value — this will allow for future expansion in the parser modes, such as the "use / as the option character" mode that people from Windows backgrounds often ask for.

**Comment by rmunn** [comment]
@derrickstolee The 2.9 release was held up by the old NuGet API key for the package having expired. #733 updates the NuGet API key, so once that gets merged the release will be possible. Watch #733 if you want to find out when that happens. Unfortunately, I can't do anything about the release timing, as I'm not one of the project maintainers. I believe @ericnewton76 is the one who manages releases right now; Eric, please correct me if that's mistaken.

**Comment by rmunn** [comment]
#735 and #709 are also fixed by this PR, and should be closed once 2.9.0 is finally released.

**Pull request merged**
